cheatsheets_draft/GraphQL_Cheat_Sheet.md

```diff
@@ -0,0 +1,257 @@
+
+## Best Practices and Recommendations
+
+### Input Validation
+
+Adding strict input validation can help prevent against injection and DoS. The main design for GraphQL is that the user supplies one or more identifiers and the backend has a number of data fetchers making HTTP, DB, or other calls using the given identifiers. This means that user input will be included in HTTP requests, DB queries, or other requests/calls which provides opportunity for injection that could lead to various injection attacks or DoS.
+
+See the OWASP Cheat Sheets on [Input Validation](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html) and general [injection prevention](https://cheatsheetseries.owasp.org/cheatsheets/Injection_Prevention_Cheat_Sheet.html) for full details to best perform input validation and prevent injection.
+
+#### General Practices
+
+Validate all incoming data to only allow valid values (i.e. whitelist).
+
+- Use specific GraphQL [data types](https://graphql.org/learn/schema/#type-language) such as [scalars](https://graphql.org/learn/schema/#scalar-types) or [enums](https://graphql.org/learn/schema/#enumeration-types). Write custom GraphQL [validators](https://graphql.org/learn/validation/) for more complex validations. [Custom scalars](https://itnext.io/custom-scalars-in-graphql-9c26f43133f3) may also come in handy.
+- Define [schemas for mutations input](https://graphql.org/learn/schema/#input-types).
+- [Whitelist allowed characters](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#whitelisting-vs-blacklisting) - don't use a blacklist
+    - The stricter the whitelist the better. A lot of times a good starting point is only allowing alphanumeric, non-unicode characters because it will disallow many attacks.
+- To properly handle unicode input, use a [single internal character encoding](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#validating-free-form-unicode-text)
+- Gracefully [reject invalid input](https://cheatsheetseries.owasp.org/cheatsheets/Error_Handling_Cheat_Sheet.html), being careful not to reveal excessive information about how the API and its validation works.
+
+#### Injection Prevention
+
+When handling input meant to be passed to another interpreter (e.g. SQL/NoSQL/ORM, OS, LDAP, XML):
+
+- Always prefer safe tools with support for parameterized statements
+    - Ensure that you follow the documentation so you are properly using the tool
+    - For SQL, using ORMs/ORDs is a good option but they must be used properly to avoid [ORM injection](https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/07-Input_Validation_Testing/05.7-Testing_for_ORM_Injection).
```

Comment:
Is there anything better within OWASP that we can link to? The audience for WTSG isn't developers like this CS is, so another resource would work better here if it exists. I couldn't find anything.

Reply:
I didn't find anything better then the WTSG.

We could link the [CAPEC-109: Object Relational Mapping Injection](https://capec.mitre.org/data/definitions/109.html) but I like the WTSG better since it includes examples. I am OK linking the WTSG.

Reply:
Okay we'll stick with the WSTG doc then.